src/Compilers/Core/Rebuild/CSharpCompilationFactory.cs

```diff
@@ -0,0 +1,123 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.IO;
+using Microsoft.Cci;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.Text;
+using CS = Microsoft.CodeAnalysis.CSharp;
+
+namespace BuildValidator
+{
+    public sealed class CSharpCompilationFactory : CompilationFactory
+    {
+        public new CSharpParseOptions ParseOptions { get; }
```

Comment:
I have an dislike for casting like this because it makes me feel like I've done something wrong in my API design. I get the reason you want to do this though. Let me make the base `abstract` and then can just `override` here and get the single point of storage without the casting.

Reply:
Okay think I have a good solution to this now.

Reply:
It looks like the consumer of `CSharpCompilationFactory` will see two properties: `ParseOptions` and `CSharpParseOptions`, and the `CSharp` prefix seems redundant.

Could we use the same pattern as `Compilation.Options`?

```csharp
abstract class Compilation
{
    public CompilationOptions CompilationOptions => CommonOptions;
    protected abstract CompilationOptions CommonOptions { get; }
}

sealed class CSharpCompilation : Compilation
{
    private readonly CSharpCompilationOptions _options;
    public new CSharpCompilationOptions CompilationOptions => _options;
    protected override CompilationOptions CommonOptions => _options;
}
```

---
In reply to: [600039774](https://github.com/dotnet/roslyn/pull/52079#discussion_r600039774) [](ancestors = 600039774)